### PR TITLE
Add resource quota for ConfigMap

### PR DESCRIPTION
**Pull request by pmorie**
...and fix a little bug for resource quota for secret.

@derekwaynecarr

### CODE REVIEW

File: test/e2e/resource_quota.go [hunk 1]
```diff
@@ -342,6 +414,8 @@ func newTestResourceQuota(name string) *api.ResourceQuota {
 	hard[api.ResourceQuotas] = resource.MustParse("1")
 	hard[api.ResourceCPU] = resource.MustParse("1")
 	hard[api.ResourceMemory] = resource.MustParse("500Mi")
+	hard[api.ResourceConfigMaps] = resource.MustParse("2")
+	hard[api.ResourceSecrets] = resource.MustParse("2")
```
Comment (derekwaynecarr) [inline test/e2e/resource_quota.go]: maybe comment here as well why you set this to 2...

**Comment by derekwaynecarr** [comment]
code lgtm except for the minor nit to add a comment...

@bgrant0607 - please ack on the resource name used.

**Comment by derekwaynecarr** [comment]
actually, you need to add configmaps to this list:
https://github.com/pmorie/kubernetes/blob/config-quota/pkg/api/helpers.go#L180

**Comment by bgrant0607** [comment]
cc @mqliang

**Comment by bgrant0607** [comment]
@derekwaynecarr The name looks consistent with the pattern of the others. Did you have a specific concern? If not, then Ack.

**Comment by derekwaynecarr** [comment]
@bgrant0607 - no concern, just fyi .

**Comment by derekwaynecarr** [comment]
Thanks @pmorie !

**Comment by pmorie** [comment]
@derekwaynecarr the pleasure was mine, sir

**Comment by pmorie** [comment]
@derekwaynecarr rebased, glad we fixed that other flake

**Comment by pmorie** [comment]
@k8s-bot test this issue: #21753

**Comment by pmorie** [comment]
@k8s-bot test this issue: #21244

File: test/e2e/resource_quota.go (hunk 1 above)
Comment (derekwaynecarr) [inline test/e2e/resource_quota.go]: Remove the defaulting to 2 for secrets.

**Comment by derekwaynecarr** [comment]
@pmorie remove the line on default secret quota to 2 and then this is Lgtm

**Comment by thockin** [comment]
I did not realize we had quota for types like this.  Why?  It's another place where we have the modularity inside out, and types have debris scattered all over.

**Comment by derekwaynecarr** [comment]
@thockin - are you asking why you want to quota the number of config maps in a namespace, or why doing that requires code?  If its the latter, I agree that I want to avoid having to write code specific to counting objects, and instead move generic object count quota to be driven based off discovery and the dynamic client if all you are interested in quota on that object is the count itself.  Both discovery and the dynamic client came in super late to kube 1.2 so I have not had time to adjust, but agreed they are necessary next steps.   That said, you still need a name to address a type by in the quota document, and I think we want that name to be independent of version.

We will always need some code in quota to calculate more complex things that are derived from a kind.  For example, the quota calculation around pods is more complex since it carries dependent resource concepts (requests/limits for cpu, memory, etc.).  

see: https://github.com/kubernetes/kubernetes/blob/master/pkg/quota/evaluator/core/pods.go#L35

Similarly, I would expect more complicated semantics around persistent volume claims (where i will want to quota sum of pvc requests in a namespace, and sum of bound pv sizes to pvcs).  In addition, for services, I want to quota the number of load balancers, and node ports they may consume.  Downstream in OpenShift, we quota images and image streams for the size of the image in addition to the counts.

To summarize, many objects actually quota more things than just the generic count, but if you just want to do generic count, I do want to move to either discovery/dynamic client, and/or code generation.

**Comment by derekwaynecarr** [comment]
@pmorie - I think we can fix the hard-coded e2e secret count in a follow-on.

**Comment by thockin** [comment]
OK, but why quota config maps?

**Comment by derekwaynecarr** [comment]
@thockin - any of the following:

* a cluster-admin wants to stop users from causing unbounded memory growth in etcd; right now, everything in etcd is held in memory with about a 13-18:1 ratio of in-memory versus on-disk cost depending on the resource, so object counts generally across the cluster are a concern when you offer a cluster to a set of untrusted users, and you want to protect yourself as the operator.
* cluster-admins that manage kubernetes with differentiated levels of service for others, for example, a free tier gets 10 config maps, a paid tier gets 100 type of thing...

**Comment by thockin** [comment]
Any object the user can create has labels and annotations, and likely other
insufficiently-validate or free-form fields.  Quota-ing the number of
utility objects a user can create seems like putting out a fire with a
thimble of water.  Objects like ConfigMap don't even have a real
manifestation like a pod.

This is a relatively minor point, but I see VERY little value in making
things like ConfigMap be an up-sell.  It's just going to annoy people when
they hit limits, but not actually provide any safety or additional value.
Conversely, I see a real cost - PRs like this that accumulate (IMO)
pointless code.

Up-sell number of pods, or services, or total resource footprint.
ConfigMap is a utility.

I'll go back to my cave now.

**Comment by derekwaynecarr** [comment]
@thockin - I am not sure what to say.... we have had entire clusters go down because of uncontrolled object proliferation/spam bringing down etcd, and we do have customers that want to control for that scenario.  So ignore the upsell scenario, that is probably more interesting for other resource types, but spam is a problem...

**Comment by thockin** [comment]
If spam is a problem we should be looking at heuristics that control
something meaningful.  Total objects, footprint, something more meaningful
than how many config maps you can create.  I suppose if this was
encapsulated and not open-coded I wouldn't hate it so much.  Sigh.

Anyway, there's precedent for this, so I'll just go away, but I think it's
silly, personally.

**Comment by derekwaynecarr** [comment]
@thockin - fair enough, I am all for deprecating things if we end up with better solutions in the future.
